File: osp/osp.py
import numpy as np
from numpy import linalg as LA
from  scipy import optimize 
import scipy as sp
import sys



def scaled_chebyshev_basis(s,p,zmin,zmax,z):
  b = np.zeros((s+1,p+1))
  m1 = 2/(zmax-zmin)
  m0 = -(1+zmin*m1);

  # T_0 = 1
  b[0][0] = 1;

  # T_1 = m1*x + m0
  b[1][0] = m0;
  b[1][1] = m1;

  # T_{n+1} = 2*(m1*x + m0)*T_{n} - T_{n-1}
  # L_{n} = (2n-1)/n*(m1*x+m0)*L_{n} - (n-1)/n*L_{n-1}
  zero_arr = np.zeros(1)
  for k in range(0,s-1):
#    b[k+2][:] = (2.0*(k+2)-1)/(k+2)*(m1*np.concatenate((zero_arr,b[k+1][0:-1])) + m0*b[k+1][:]) - (k+2-1.0)/(k+2)*b[k][:]
    b[k+2][:] = 2*(m1*np.concatenate((zero_arr,b[k+1][0:-1])) + m0*b[k+1][:]) - b[k][:]

  c= np.zeros((s+1,len(z)),dtype=z.dtype)
  c[0][:] = 1;
  c[1][:] = m1*z+m0;

  for k in range(0,s-1):
#    c[k+2][:] = (2.0*(k+2)-1)/(k+2)*(m1*z + m0)*c[k+1][:] - (k+2.0-1)/(k+2)*c[k][:]
    c[k+2][:] = 2*(m1*z + m0)*c[k+1][:] - c[k][:]

  return [b,c]

def minimizePoly(s,p,h,ev_space,eta,tol,maxiter=128,verbose=False,poly_guess=None):
  if verbose:
    print "============================================="
  hval  = h*np.real(ev_space) + 1j*np.imag(ev_space)
  hval = h*ev_space;
  [b,c] = scaled_chebyshev_basis(s,p,min(np.real(hval)),0,hval)


  fixed_coefficients = np.ones(p+1)/sp.misc.factorial(np.linspace(0,p,p+1))

  def func(x):
    return x[-1];
  def func_jac(x):
    return np.concatenate((np.zeros(len(x)-1),[1]))

  def cfunc1(x,b,coeff):
    return np.dot(x[:-1],b) - coeff
  def cfunc1_jac(x,J):
    return J;

  if True:
    def cfunc2(x,c):
      g = np.dot(x[:-1],c);
      f = np.real(g*np.conj(g)) - 1
      imax = np.argmax(f)
      return np.real(x[-1] - f[imax])
    def cfunc2_jac(x,c):
      g  = np.dot(x[:-1],c[:-1]);
      f = np.real(g*np.conj(g)) - 1
      imax = np.argmax(f)
      ct = c.T;
      df = -ct[imax]*np.conj(g[imax]);
      df[-1] = 0.5;
      return np.real(df + np.conj(df)).T
  else:
    def cfunc2(x,c):
      g = np.dot(x[:-1],c);
      return np.real(x[-1]*np.ones(len(g)) - (g*np.conj(g) - 1))
    def cfunc2_jac(x,c):
      g  = np.dot(x[:-1],c[:-1]);
      df = -c*np.conj(g);
      df[-1] = 0.5;
      return np.real(df + np.conj(df)).T


  m,n = b.shape
  m,k = c.shape
  J1 = np.zeros((s+2,n))
  J2 = np.zeros((s+2,k),dtype=c.dtype)

  for i in range(s+1):
    J1[i] = b[i]
    J2[i] = c[i]
  J1 = J1.T

  x0 = np.zeros(s+2);
  cons = ({'type': 'eq',
    'fun' : lambda x: cfunc1(x,b,fixed_coefficients),
    'jac' : lambda x: cfunc1_jac(x,J1)},
    {'type': 'ineq',
    'fun' : lambda x: cfunc2(x,c)})
  #  'jac' : lambda x: cfunc2_jac(x,J2)})
    
  res=optimize.minimize(func, x0, constraints=cons, jac=func_jac,
      method='SLSQP', options={'disp': verbose, 'maxiter': maxiter}, tol=1e-13)

  if verbose:
    print "------------------------------------"
    print res
    print "------------------------------------"
    print 'Value= ', res.fun-eta
    print "coeff= "
    for x in np.dot(res.x[:-1],b):
      sys.stdout.write("%.16g," % x)
    print ""
    for x in fixed_coefficients:
      sys.stdout.write("%.16g," % x)
    print "\n============================================="


  if res.success:
    return [True, res.x[:-1], res.fun, res.nit]
  elif abs(res.fun-eta) < tol:
    return [True, res.x[:-1], res.fun, res.nit]
  else:
    return [False, res.x, res.fun, res.nit]

def maximizeH(s,p,ev_space):
  h_min = 0; #60 #0.00*max(abs(ev_space))
  h_max = 2.01*s*s*max(abs(ev_space))

  max_iter = 1280;
  max_steps = 1000
  tol_bisect = 0.001
  tol_feasible = 1.0e-12
  eta = 0.0

  print "max_iter= %d " % max_iter
  print "max_steps= %d " % max_steps
  print "tol_bisect= %g " % tol_bisect
  print "tol_feasible= %g " % tol_feasible
  print "eta= %g" % eta

  poly = None
  v    = None
  converged = False;
  for step in range(max_steps):
    if ((h_max-h_min < tol_bisect*h_min) or (h_max < tol_bisect)):
      if converged:
        break;
      else:
        h = h_min
    else:
      h = 0.25*h_max + 0.75*h_min
      h = 0.5*h_max + 0.5*h_min


#    h = 0.5*h_max + 0.5*h_min

    [conv, poly, v, nit] = minimizePoly(s,p,h,ev_space,eta,tol_feasible,max_iter,verbose=False)
    print "%5d  h_min= %g   h_max= %g  -- h= %g nit= %d  v= %g " % (step, h_min, h_max, h, nit, v-eta)

    if False:
      if not conv:
        converged = False
        print " >>>> Failed to converge "
        h_max = h;
      else:
        converged = True
        if abs(v-eta) <= tol_feasible:
          h_min = h;
        else:
          h_max = h;
    else:
      if not conv or abs(v-eta) > tol_feasible:
        converged = False
        print " >>>> Failed to converge "
        h_max = h;
      else:
        converged = True
        h_min = h;


  if True or converged:
    print " Converged with h= %g  h/s^2= %g" % (h, h/s**2)
    [conv, poly, v, nit] = minimizePoly(s,p,h,ev_space,eta,tol_feasible,max_iter,verbose=True)
    return [poly, h]
  else:
    return [None, None]


if False:
  npts = 1000;

  ev_space = -np.linspace(0,1,npts);

#  ev_space = np.random.random(npts);
#  ev_space = np.sort(np.concatenate(([0],ev_space,[1])))
#  ev_space = -0.5*(1 + np.cos(ev_space*np.pi))

#  ev_space = np.linspace(0,np.pi,npts)
#  ev_space = -0.5*(1 + np.cos(ev_space))
  ev_space = -np.linspace(0,1,npts);

  if True:
    kappa=1;
#    beta =5.0;
    beta = 0.2;

    imag_lim = beta;
    l1 = 1j*np.linspace(0,imag_lim,npts/2);
    l2 = 1j*imag_lim + np.linspace(-kappa,0,npts);
    l3 = -kappa + 1j*np.linspace(0,imag_lim,npts/2);
    ev_space = np.concatenate((l1,l2,l3));


  s = 30;
  p = 8;

  print "npts= %d  s= %d  p= %d " % (npts, s, p)

  [poly, h] = maximizeH(s,p,ev_space);
  if h != None:
    print "------ Polynomial coefficients -------- "
    print "coeff= {"
    for x in poly[:-1]:
      sys.stdout.write("%.16g," % x)
    sys.stdout.write("%.16g};\n" % poly[-1])
    print "h= %.16g  h/s^2= %g " % (h, h/(s*s))
  else:
    print " ------- Not converged ------ "

  sys.exit(-1)




if True:
  npts = 1000;
  kappa=1;
  beta = 0.5;
  beta= 0.5;

  imag_lim = beta;
  l1 = 1j*np.linspace(0,imag_lim,50);
  l2 = 1j*imag_lim + np.linspace(-kappa,0,npts);
  l3 = -kappa + l1 #1j*np.linspace(0,imag_lim,50);
  ev_space = np.concatenate((l1,l2,l3));
#  ev_space =  np.linspace(-kappa,0,npts);
  eta=0
  tol=1.0e-12
  h=35.58

  p=4
  s=30

  maxiter=1024
  [conv, poly, v, nit] = minimizePoly(s,p,h,ev_space,eta,tol,maxiter,verbose=True)

  if not conv:
    print " ------- Not converged ------ "

  print "------ Polynomial coefficients -------- "
  print "coeff= {"
  for x in poly[:-1]:
    sys.stdout.write("%.16g," % x)
  sys.stdout.write("%.16g};\n" % poly[-1])
  print "h= %.16g  h/s^2= %g " % (h, h/(s*s))

File: osp/osp.cpp
#include <iostream>
#include <nlopt.hpp>
#include <complex>
#include <cmath>
#include <algorithm>
#include <numeric>
#include <cassert>
#include "common.h"

using std::get;


template<typename real_type, typename complex_type>
static auto scaled_chebyshev_basis(
    const size_t s, const size_t p,
    const real_type zmin,
    const real_type zmax,
    const std::vector<complex_type>& z)
{
  const auto m1 = 2/(zmax - zmin);
  const auto m0 = -(1 + 0.000 + zmin*m1);

  std::vector<real_type> _b((s+2)*(p+1),0);
  {
    auto b = [&](const size_t _s, const size_t _p) -> real_type&
    {
      return _b[_p*(s+2) + _s];
    };
    b(0,0) = 1;
    b(1,0) = m0;
    b(1,1) = m1;
    for (size_t k = 0; k < s-1; k++)
    {
      b(k+2,0) = 2.0*(0 + m0*b(k+1,0)) - b(k,0);
      for (size_t j = 1; j < p+1; j++)
      {
        b(k+2,j) = 2.0*(m1*b(k+1,j-1) + m0*b(k+1,j)) - b(k,j);
      }
    }
  }

  std::vector<complex_type> _c((s+2)*z.size(),0);
  {
    auto c = [&](const size_t _s, const size_t _i) -> complex_type&
    {
      return _c[_i*(s+2) + _s];
    };
    for (size_t j = 0; j < z.size(); j++)
    {
      c(0,j) = 1;
      c(1,j) = m1*z[j] + m0;
    }
    for (size_t k = 0; k < s-1; k++)
      for (size_t i = 0; i < z.size(); i++)
      {
        c(k+2,i) = 2.0*(m1*z[i]+m0)*c(k+1,i) - c(k,i);
      }
  }

  return std::make_tuple(_b, _c);
}

template<typename real_type, typename complex_type>
auto optimize(const size_t p, const size_t s, const real_type beta_p, const real_type h_scale, const std::vector<complex_type> &ev_space, const bool verbose = true, const bool verbose_exception = true, const bool sanity_check = true)
{
  auto h_space = ev_space;
  real_type h_min = 0;
  real_type h_max = 0;

  for (auto& h : h_space)
  {
    using namespace std::literals;
    const auto h0 = h;
    h = h0*h_scale;
    h = h0.real()*h_scale + 1i*h0.imag()*std::pow(h_scale,beta_p);
    h_min = std::min(h_min, h.real());
  }
  assert(h_min < h_max);
  
  auto basis = scaled_chebyshev_basis(s,p,h_min,h_max,h_space);

 
  std::vector<real_type> fixed_coeff(p+1);
  real_type factorial = 1;
  for (size_t i = 0; i < p+1; i++)
  {
    factorial *= std::max(size_t{1},i);
    fixed_coeff[i] = 1.0/factorial;
  }

  using std::get;
  struct func_eq_data_type
  {
    std::vector<real_type>* bmat_ptr, *coeff_ptr;
    size_t p;
    size_t s;
  } func_eq_data;
  func_eq_data.bmat_ptr  = &get<0>(basis);
  func_eq_data.coeff_ptr = &fixed_coeff;
  func_eq_data.p         = p;
  func_eq_data.s         = s;

  struct func_ineq_data_type
  {
    std::vector<complex_type> *cmat_ptr;
    size_t npts;
    size_t s;
  } func_ineq_data;
  func_ineq_data.cmat_ptr = &get<1>(basis);
  func_ineq_data.npts     = h_space.size();
  func_ineq_data.s        = s;

  nlopt::func func = [](
      unsigned n, const double *x,
      double *grad,
      void *func_data)
  {
    if (grad)
    {
      std::fill(grad, grad+n-1, 0);
      grad[n-1] = 1;
    }
    return x[n-1];
  };

  nlopt::mfunc func_eq = [](
      unsigned m, double *result,
      unsigned n, const double *x,
      double *grad, /* NULL if not needed */
      void *func_data)
  {
    assert(func_data);
    const auto &data  = *reinterpret_cast<func_eq_data_type*>(func_data);
    const auto& bmat  = *data.bmat_ptr;
    const auto& coeff = *data.coeff_ptr;
    const auto s = data.s;
    const auto p = data.p;
    assert(m == p+1);
    assert(n == s+2);
    assert(m*n == bmat.size());
    if (grad)
    {
      std::copy(bmat.begin(), bmat.end(), grad);
    }
    for (size_t i = 0; i < m; i++)
    {
      result[i] = -coeff[i] +
        std::inner_product(x, x+n-1, bmat.begin() + i*n, 0.0);
    }
  };

  auto n_ineq = h_space.size();
  nlopt::mfunc func_ineq = [](
      unsigned m, double *result,
      unsigned n, const double *x,
      double *grad, /* NULL if not needed */
      void *func_data)
  {
    assert(func_data);
    const auto &data = *reinterpret_cast<func_ineq_data_type*>(func_data);
    const auto& cmat = *data.cmat_ptr;
    const auto s = data.s;
    assert(m == data.npts);
    assert(n == s+2);
    using std::conj;
    using std::real;
    if (grad)
    {
      for (size_t i = 0; i < m; i++)
      {
        const auto g = 
          std::inner_product(x, x+n-1, cmat.begin() + i*n,complex_type{0});
        for (size_t j = 0; j < s+1; j++)
        {
          const auto df = cmat[i*n+j]*conj(g);
          grad[i*n+j] = real(df + conj(df));
        }
        grad[i*n+s+1] = -1;

        const auto re = real(g*conj(g));
        result[i] = (re-1)-x[n-1];
      }
    }
    else
    {
      for (size_t i = 0; i < m; i++)
      {
        const auto g = 
          std::inner_product(x, x+n-1, cmat.begin() + i*n,complex_type{0});
        const auto re = real(g*conj(g));
        result[i] = (re-1)-x[n-1];
      }
    }
  };
  if (0)
  {
    n_ineq = 1;
    func_ineq = [](
        unsigned m, double *result,
        unsigned n, const double *x,
        double *grad, /* NULL if not needed */
        void *func_data)
    {
      assert(func_data);
      const auto &data = *reinterpret_cast<func_ineq_data_type*>(func_data);
      const auto& cmat = *data.cmat_ptr;
      const auto s = data.s;
      assert(m == 1);
      assert(n == s+2);
      using std::conj;
      using std::real;

      real_type fmax = -1;
      real_type res  = 0;
      int       imax = 0;
      for (size_t i = 0; i < data.npts; i++)
      {
        const auto g = 
          std::inner_product(x, x+n-1, cmat.begin() + i*n,complex_type{0});
        const auto f = real(g*conj(g))-1;
        if (f > fmax)
        {
          fmax = f;
          res =  f - x[n-1];
          imax = i;
        }
      }
      result[0] = res;

      if (grad)
      {
        const auto g = 
          std::inner_product(x, x+n-1, cmat.begin() + imax*n,complex_type{0});
        for (size_t j = 0; j < s+1; j++)
        {
          const auto df = cmat[imax*n+j]*conj(g);
          grad[j] = real(df + conj(df));
        }
        grad[s+1] = -1;
      }
    };
  }



  auto method = nlopt::LD_SLSQP;
  nlopt::opt opt(method, s+2);

  opt.set_min_objective(func,NULL);

  const real_type tol = 1.0e-13;
  opt.add_equality_mconstraint(
      func_eq, 
      &func_eq_data, 
      std::vector<real_type>(p+1,tol));
  
  std::vector<real_type> x(s+2, 1);
  opt.add_inequality_mconstraint(
      func_ineq, 
      &func_ineq_data, 
      std::vector<real_type>(n_ineq,tol)); 

  opt.set_xtol_abs(tol);
//  opt.set_ftol_abs(1.0e-14);


  real_type minf;
  decltype(opt.optimize(x,minf)) result;
  try
  {
    result = opt.optimize(x,minf);
  }
  catch (std::exception &e)
  {
    if (verbose_exception)
      printf(std::cerr, " Exception caught: % \n", e.what());
  }

  if (verbose)
  {
    printf(std::cerr, "result= % \n", result);
    printf(std::cerr, "minf= % \n", minf);
  }

  if (sanity_check)
  {
    const auto &poly = x;
    const auto &bmat = get<0>(basis);
    std::cerr << "fixed_coeff_comp= " << std::setprecision(16);
    for (int i = 0; i < p+1; i++)
    {
      const auto res = std::inner_product(poly.begin(), poly.end()-1, bmat.begin() + i*(s+2), 0.0);
      std::cerr << res << " ";
    }
    std::cerr << std::endl;
    
    std::cerr << "fixed_coeff_gold= " ;
    for (int i = 0; i < p+1; i++)
    {
      const auto res = fixed_coeff[i];
      std::cerr << res << " ";
    }
    std::cerr << std::endl;
    
    std::cerr << "fixed_coeff_diff= " ;
    for (int i = 0; i < p+1; i++)
    {
      const auto res = std::inner_product(poly.begin(), poly.end()-1, bmat.begin() + i*(s+2), 0.0);
      std::cerr << std::abs(res-fixed_coeff[i])/fixed_coeff[i] << " ";
    }
    std::cerr << std::endl;
    std::cerr << " ===================================== \n";


  }

  return std::make_tuple(
      std::vector<real_type>(x.begin(), x.end()-1),
      minf);
}

template<typename real_type>
std::vector<real_type> linspace(const real_type a, const real_type b, const size_t n)
{
  std::vector<real_type> res(n);
  std::iota(res.begin(), res.end(), 0);
  for (auto& x : res)
    x = a + (b-a)*x/(n-1); 
  return res;
}

template<typename real_type, typename complex_type>
auto maximizeH(const size_t p, const size_t s, const real_type beta_p, const std::vector<complex_type>& ev_space)
{

  real_type h_min = 0; 
  real_type h_max = 0;
  for (auto &x : ev_space)
    h_max = std::max(h_max, std::abs(std::real(x)));
  h_max *= 2.01*s*s/p;

//  const auto max_iter = 1280;
  const auto max_steps = 1000;
  const auto tol_bisect = 0.01;

//  printf(std::cerr, "max_iter= % \n", max_iter);
  printf(std::cerr, "max_step= % \n", max_steps);
  printf(std::cerr, "tol_bisct= % \n", tol_bisect);

  bool converged = false;
  auto h = h_min;
  printf(std::cerr, " -- begin-loop -- \n");
  for (auto step = 0; step < max_steps; step++)
  {
    if ((h_max-h_min < tol_bisect*h_min) or (h_max < tol_bisect))
    {
      if (converged)
        break;
      else
        h = h_min;
    }
    else
    {
      h = 0.25*h_max + 0.75*h_min;
    }

    const auto& res = optimize(p, s, beta_p, h, ev_space, false, false,false);
    printf(std::cerr, "step= %  h_min= %  h_max= %  -- h= %  val= % \n",
        step, h_min, h_max, h, get<1>(res));

    if (std::abs(get<1>(res)) < 1.0e-12)
    {
      converged = true;
      h_min = h;
    }
    else
    {
      converged = false;
      h_max = h;
    }
  }

  assert(converged);
  const auto& res = optimize(p, s, beta_p, h, ev_space);
  printf(std::cerr, "step= %  h_min= %  h_max= %  -- h= %  val= % \n",
      -1, h_min, h_max, h, get<1>(res));

  return std::make_tuple(get<0>(res), h);
}

void test()
{
  using namespace std::literals;
  using real_type    =  double;
  using complex_type = std::complex<real_type>;

  const size_t npts = 1000;

  const real_type kappa  = 1;
  const real_type beta   = 0.5;
  const real_type beta_p = 0.25;


  const auto imag_lim = std::abs(beta);
  const auto l1 = linspace(0.0,imag_lim,npts);
  const auto l2 = linspace(-kappa,0.0,npts);

  std::vector<complex_type> ev_space;
  if (imag_lim > 0)
    for (auto& x : l1)
      ev_space.emplace_back(1i*x);
  for (auto& x : l2)
    ev_space.emplace_back(1i*imag_lim + x);
  if (imag_lim > 0)
    for (auto &x : l1)
      ev_space.emplace_back(-kappa + 1i*x);

  size_t p = 4;
  size_t s = 30;
  real_type h = 35.58;


  const auto res = std::get<0>(optimize(p, s, beta_p, h, ev_space));
  std::cout << "Coefficients: \n";
  for (auto & x : res)
  {
    std::cout << x << ", ";
  }
  std::cout << std::endl;
  {
    std::cerr << " ----------- \n";
    const auto res = std::get<0>(optimize(p, s, beta_p, h, ev_space));
    std::cout << "Coefficients: \n";
    for (auto & x : res)
    {
      std::cout << x << ", ";
    }
    std::cout << std::endl;
  }
}

void maximizeHdriver()
{
  using namespace std::literals;
  using real_type    =  double;
  using complex_type = std::complex<real_type>;

  int s = 30;
  int p = 8;

  size_t npts = 1000;
  real_type kappa  = 1;
  real_type beta   = 1; //1/10;
  real_type beta_p = 0.25;
  beta = 1.0/10;
  beta = 1.0e-14;
  beta= 0.1;

  beta_p = 0.3;
  s = 100;
//  s = std::min(s,std::max(p,static_cast<int>(std::sqrt(1.0/beta/0.15))+1));
//
  beta = 1.0;

  beta_p = 0.35;
  beta_p = 1;
  beta   = 1.0/100;

  s      = 110;
  real_type zeta = 1.0/std::sqrt(0.15);
  real_type stiff_kappa = 1000;
  s = std::max(p,static_cast<int>(zeta*std::sqrt(stiff_kappa) + 1));
#if 0 /* basic */
  beta = 1.0/stiff_kappa;
  beta_p = 1.0;
#else
  beta_p = 0.3;
  beta   = 1.0;
#endif


#if 0
  beta = 0.5;
  s    = 12;
#endif

  const auto imag_lim = std::abs(beta);
  const auto l1 = linspace(0.0,imag_lim,npts/2);
  const auto l2 = linspace(-kappa,0.0,npts);

  std::vector<complex_type> ev_space;
  if (imag_lim > 0)
    for (auto& x : l1)
      ev_space.emplace_back(1i*x);
  for (auto& x : l2)
    ev_space.emplace_back(1i*imag_lim + x);
  if (imag_lim > 0)
    for (auto &x : l1)
      ev_space.emplace_back(-kappa + 1i*x);
  
  printf(std::cerr, "npts= % \n", npts);
  printf(std::cerr, "p= % \n", p);
  printf(std::cerr, "s= % \n", s);
  printf(std::cerr, "kappa= % \n", kappa);
  printf(std::cerr, "kappa_stiff= % \n", stiff_kappa);
  printf(std::cerr, "beta= % \n",  beta);
  printf(std::cerr, "beta_p= % \n",  beta_p);


  const auto res = maximizeH<real_type,complex_type>(p,s,beta_p,ev_space);


  std::cout << "coeff = { \n";
  const auto & poly = get<0>(res);
  const auto & h    = get<1>(res);
  for (size_t i = 0; i < poly.size() -1 ; i++)
  {
    std::cout << std::setprecision(16) << poly[i] << ", ";
  }
  std::cout << poly.back()  << " };\n";
  std::cout << "h= " << h << std::endl;
  std::cout << "h/s^2= " << h/(s*s) << std::endl;
  std::cout << "h_imag= " << std::pow(h,beta_p)*beta << std::endl;
  std::cout << "h_imag/s= " << std::pow(h,beta_p)*beta/s << std::endl;
}


int main(int argc, char * argv[])
{
#if 0
  test();
#else
  maximizeHdriver();
#endif
  return 0;
}
